gio/src/datagram_based.rs

```diff
@@ -0,0 +1,266 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::Cancellable;
+use crate::DatagramBased;
+use crate::InputMessage;
+use crate::OutputMessage;
+use futures_core::stream::Stream;
+use glib::object::{Cast, IsA};
+use glib::translate::*;
+use std::cell::RefCell;
+use std::mem::transmute;
+use std::pin::Pin;
+use std::ptr;
+use std::time::Duration;
+
+pub trait DatagramBasedExtManual: Sized {
+    #[doc(alias = "g_datagram_based_create_source")]
+    fn create_source<F, C>(
+        &self,
+        condition: glib::IOCondition,
+        cancellable: Option<&C>,
+        name: Option<&str>,
+        priority: glib::Priority,
+        func: F,
+    ) -> glib::Source
+    where
+        F: FnMut(&Self, glib::IOCondition) -> glib::Continue + 'static,
+        C: IsA<Cancellable>;
+
+    fn create_source_future<C: IsA<Cancellable>>(
+        &self,
+        condition: glib::IOCondition,
+        cancellable: Option<&C>,
+        priority: glib::Priority,
+    ) -> Pin<Box<dyn std::future::Future<Output = glib::IOCondition> + 'static>>;
+
+    fn create_source_stream<C: IsA<Cancellable>>(
+        &self,
+        condition: glib::IOCondition,
+        cancellable: Option<&C>,
+        priority: glib::Priority,
+    ) -> Pin<Box<dyn Stream<Item = glib::IOCondition> + 'static>>;
+
+    #[doc(alias = "g_datagram_based_condition_wait")]
+    fn condition_wait(
+        &self,
+        condition: glib::IOCondition,
+        timeout: Option<Duration>,
+        cancellable: Option<&impl IsA<Cancellable>>,
+    ) -> Result<(), glib::Error>;
+
+    #[doc(alias = "g_datagram_based_receive_messages")]
+    fn receive_messages<'v, V: IntoIterator<Item = &'v mut [&'v mut [u8]]>, C: IsA<Cancellable>>(
+        &self,
+        messages: &mut [InputMessage],
+        flags: i32,
+        timeout: Option<Duration>,
+        cancellable: Option<&C>,
+    ) -> Result<usize, glib::Error>;
+
+    #[doc(alias = "g_datagram_based_send_messages")]
+    fn send_messages<C: IsA<Cancellable>>(
+        &self,
+        messages: &mut [OutputMessage],
```

Comment:
Why is this mutable?

Reply:
The `bytes_sent` field is modified by this function

Reply:
Ah that makes sense